packages/controls/src/controls/slot/slot.ts

```diff
@@ -0,0 +1,164 @@
+import {
+  type CopyContext,
+  type MergeContext,
+  type MergeTranslatableDataContext,
+} from '../../context'
+import { Targets, type IntrospectionTarget } from '../../introspection'
+import { type SerializedRecord } from '../../serialization'
+
+import { ControlDefinition, serialize, type SchemaType } from '../definition'
+import { type SendMessage } from '../instance'
+
+import { SlotControl } from './slot-control'
+
+type Schema = typeof Definition.schema
+type DataType = z.infer<Schema['data']>
+type ValueType = z.infer<Schema['value']>
+
+abstract class Definition<RuntimeNode> extends ControlDefinition<
+  typeof Definition.type,
+  unknown,
+  DataType,
+  ValueType,
+  RuntimeNode,
+  SlotControl
+> {
+  static readonly type = 'makeswift::controls::slot' as const
+
+  static get schema() {
+    const type = z.literal(this.type)
+    const columnData = z.object({
+      count: z.number(),
+      spans: z.array(z.array(z.number())),
+    })
+
+    const element = Schema.element.and(
+      z.object({ deleted: z.boolean().optional() }),
+    )
+
+    const data = z.object({
+      elements: z.array(element),
+      columns: Schema.responsiveValue(columnData),
+    })
+
+    const value = data
+
+    const definition = z.object({
+      type,
+    })
+
+    return {
+      type,
+      data,
+      value,
+      definition,
+    }
+  }
+
+  constructor() {
+    super({})
+  }
+
+  get controlType() {
+    return Definition.type
+  }
+
+  get schema() {
+    return {
+      ...Definition.schema,
+      resolvedValue: this.nodeSchema,
+    }
+  }
+
+  get nodeSchema(): SchemaType<RuntimeNode> {
+    return z.any()
+  }
+
+  safeParse(data: unknown | undefined): ParseResult<DataType | undefined> {
+    return safeParse(this.schema.data, data)
+  }
+
+  fromData(data: DataType | undefined): ValueType | undefined {
+    return data
+  }
+
+  toData(value: ValueType): DataType {
+    return value
+  }
+
+  copyData(
+    data: DataType | undefined,
+    context: CopyContext,
+  ): DataType | undefined {
+    if (data == null) return data
+
+    return {
+      ...data,
+      elements: data.elements.map((element) => context.copyElement(element)),
+    }
+  }
+
+  mergeData(
+    base: DataType | undefined = { columns: [], elements: [] },
+    override: DataType | undefined = { columns: [], elements: [] },
+    context: MergeContext,
+  ): DataType {
+    const mergedColumns = base.columns
+    const mergedElements = base.elements.flatMap((baseElement) => {
+      const overrideElement = override.elements.find(
+        (e) => baseElement.type === e.type && baseElement.key === e.key,
+      )
+
+      if (overrideElement == null) return [baseElement]
+      if (overrideElement.deleted) return []
+      if (isElementReference(overrideElement)) return [overrideElement]
+      if (isElementReference(baseElement)) return [baseElement]
+
+      return context.mergeElement(baseElement, overrideElement)
```

Comment:
Original is here: https://github.com/makeswift/makeswift/pull/781/files#diff-cd4a7f92cff39705fbf6faabdc3d088cc5910e7476d67c3b5ce3d28c4675f197L78